include/koa/authentication/passport-strategies/salesforce_authentication_strategy.js

```diff
@@ -0,0 +1,74 @@
+const CustomStrategy = require('passport-custom').Strategy;
+const strategyServices = require('./strategy_services');
+const request = require('request-promise');
+const Promise = require('bluebird');
+const SALESFORCE_OAUTH_TOKEN_URL = process.env.SALESFORCE_OAUTH_TOKEN_URL || '';
+const SALESFORCE_OAUTH_AUTHORIZE_URL = process.env.SALESFORCE_OAUTH_AUTHORIZE_URL || '';
+
+module.exports = (pb) => {
+    async function getSalesforceSettings(req) {
+        let pluginService = new pb.PluginService({
+            site: req.ctx.__site
+        });
+        pluginService = Promise.promisifyAll(pluginService);
+        const settings = await pluginService.getSettingsKVAsync('tn_join');
+        return settings;
+    }
+
+    async function salesforceCallback(req, done) {
+        try {
+            const loginContext = req.session._loginContext || {};
+            const state = req.url.query.state;
+            const code = req.url.query.code;
+            const settings = await getSalesforceSettings(req);
+            const options = {
+                url: `${SALESFORCE_OAUTH_TOKEN_URL}?client_id=${settings.salesforce_client_id}&redirect_uri=${settings.salesforce_callback_url}&grant_type=authorization_code&code=${code}&client_secret=${settings.salesforce_client_secret}`,
```

Comment:
I had to read this four times before it registered - Query String.  Not Ternary.  I was just so focused on how this was going to error since the ? was in the string... I need more sleep.

Reply:
hahahaha